pkg/configurations/configurations.go

```diff
@@ -0,0 +1,157 @@
+	if err != nil {
+		log.Error(err, "Failed to unmarshal config file")
+		return
+	}
+}
+
+func exists(path string) bool {
+	_, err := os.Stat(path)
+	if err == nil {
+		return true
+	}
+	if os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+func ConfigWatcher() error {
+
+	ext := filepath.Ext(configPath)
+	if !funk.Contains(viper.SupportedExts, ext[1:]) {
+		return fmt.Errorf("file extension %s is not supported", ext)
+	}
+
+	configDir, _ := filepath.Split(configPath)
+	if !exists(configDir) {
+		return fmt.Errorf("folder %s doesn't exist, can't start configuration watcher", configDir)
+	}
+
+	v.SetConfigFile(configPath)
+	v.SetTypeByDefaultValue(true)
+
+	if err := v.ReadInConfig(); err != nil { // Find and read the config file
+		log.Info("Not able to read configuration, will update default values")
+	}
+
+	c.updateConfig(*v)
+	WatchConfig(v, configPath)
+	return nil
+}
+
+//Adapted from viper WatchConfig to match nmstate configmap watch needs
+//The main changes is that there is no need to
+//preexist config file before starting the watch
+//and it will not exit on file deletion
+func WatchConfig(v *viper.Viper, fullFilePath string) {
+	initWG := sync.WaitGroup{}
+	initWG.Add(1)
+	go func() {
+		newWatcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			log.Error(err, "Failed to start fsnotify watcher")
+			return
+		}
+		defer newWatcher.Close()
+
+		configFile := filepath.Clean(fullFilePath)
+		configDir, _ := filepath.Split(configFile)
+		realConfigFile, _ := filepath.EvalSymlinks(fullFilePath)
+
+		eventsWG := sync.WaitGroup{}
+		eventsWG.Add(1)
+		go func() {
+			for {
+				select {
+				case event, ok := <-newWatcher.Events:
+					if !ok { // 'Events' channel is closed
+						eventsWG.Done()
+						return
+					}
+					currentConfigFile, _ := filepath.EvalSymlinks(fullFilePath)
+					// we only care about the config file with the following cases:
+					// 1 - if the config file was modified or created
+					// 2 - if the real path to the config file changed (eg: k8s ConfigMap replacement)
+					const writeOrCreateMask = fsnotify.Write | fsnotify.Create
+					if (filepath.Clean(event.Name) == configFile &&
+						event.Op&writeOrCreateMask != 0) ||
+						(currentConfigFile != "" && currentConfigFile != realConfigFile) {
+						realConfigFile = currentConfigFile
+						err := v.ReadInConfig()
+						if err != nil {
+							log.Error(err, "error reading config file.\n")
+						} else {
+							c.updateConfig(*v)
+						}
+					}
+				case err, ok := <-newWatcher.Errors:
+					if ok { // 'Errors' channel is not closed
+						log.Error(err, "newWatcher error\n")
+					}
+					eventsWG.Done()
+					return
+				}
+			}
+		}()
+		newWatcher.Add(configDir)
+		initWG.Done()   // done initalizing the watch in this go routine, so the parent routine can move on...
+		eventsWG.Wait() // now, wait for event loop to end in this go-routine...
+	}()
+	initWG.Wait() // make sure that the go routine above fully ended before returning
```

Comment:
Why do we need the goroutine above? You just execute it and wait for it to end, can't it be dropped?

Reply:
nope, the above goroutine is added cause of fsnotify watcher. We need to configure it and be able to run defer on if goroutine is finished. I can do it without closing fsnotify watcher but  it looks like it is not a best practice

Reply:
I'm sorry, I don't see it.  If we keep everything, but drop `go func() {` on line 115, what does it change?

Reply:
You would be able to drop the initWG too. Then you can keep the defer, it would have exactly the same scope as it has now.

Reply:
As I checked this is the way to work with fsnotify. U open routine for ents loop and wait for it.

Reply:
Example from fsnotify docs: https://godoc.org/github.com/fsnotify/fsnotify#example-NewWatcher

You indeed run a goroutine, I have nothing against the inner one (line 129). The outer one is the issue.

Reply:
But I need to run the watcher and continue on main thread without closing it. So I create go goroutine and wait in it for inner one.
I think I am missing something. But I don't see how it will work without it.

Reply:
You do initWG.Wait(), so you wait until the outer one is done. So it runs the goroutine and then it  blocks forever. You can drop it and wait for eventsWG, would do the same thing. Or maybe I need more sleep!